### PR TITLE
stake BYX + initial distribution

### PR DESCRIPTION
**Pull request by YvesToiser**
Basic Staking of BYX (no rewards)
Basic functions in all contracts (receive, fallback, withdraw)
Initial token distribution system

I have not tested this code yet. I will do a test suite this afternoon and bring corrections if needed.

### CODE REVIEW

File: truffle/contracts/StakingManager.sol [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    /**
+     * @notice withdraw stake function.
+     *
+     * @dev withdraw stake function.
+     *
+     * @param _amount the amount to withdraw from staking.
+     */
+    function withdrawStake(uint _amount) external {
+        require(BYXStaked[msg.sender] >= _amount, "Not enough BYX staked");
+        totalBYXStaked += _amount;
```
Comment (0xB19) [inline truffle/contracts/StakingManager.sol]: plutot -= non ?

File: truffle/contracts/StakingManager.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @notice deposit stake function.
+     *
+     * @dev deposit stake function.
+     *
+     * @param _amount the amount to deposit in staking.
+     */
+    function depositStake(uint _amount) external {
+        require(BYX.balanceOf(msg.sender) >= _amount, "Not enough BYX staked");
```
Comment (0xB19) [inline truffle/contracts/StakingManager.sol]: Le message d'erreur m'a un peut perturbé pour la compréhension de la ligne. J'aurais mis quelque chose comme "Not enough BYX in address balance"

File: truffle/contracts/StakingManager.sol [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+     * @notice withdraw stake function.
+     *
+     * @dev withdraw stake function.
+     *
+     * @param _amount the amount to withdraw from staking.
+     */
+    function withdrawStake(uint _amount) external {
+        require(BYXStaked[msg.sender] >= _amount, "Not enough BYX staked");
+        totalBYXStaked += _amount;
+        BYXStaked[msg.sender] += _amount;
```
Comment (0xB19) [inline truffle/contracts/StakingManager.sol]: Meme chose, -=

File: truffle/contracts/StakingManager.sol [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    /**
+     * @notice deposit stake function.
+     *
+     * @dev deposit stake function.
+     *
+     * @param _amount the amount to deposit in staking.
+     */
+    function depositStake(uint _amount) external {
```
Comment (0xB19) [inline truffle/contracts/StakingManager.sol]: On peut verifier que le montant ne soit pas 0

**Comment by YvesToiser** [comment]
Fixed
I also added a getter for the user staked amount (that will be needed by frontend
I switched the functions between an external that call the internal for gas optimisation (like openzeppellin always do)

**Pull request merged**